content/develop/concepts/app-design/multithreading.md

```diff
@@ -0,0 +1,179 @@
+
+```python
+import streamlit as st
+import time
+from threading import Thread
+
+
+class WorkerThread(Thread):
+    def __init__(self, delay):
+        super().__init__()
+        self.delay = delay
+        self.return_value = None
+
+    def run(self):
+        start_time = time.time()
+        time.sleep(self.delay)
+        end_time = time.time()
+        self.return_value = f"start: {start_time}, end: {end_time}"
+
+
+delays = [5, 4, 3, 2, 1]
+result_containers = []
+for i, delay in enumerate(delays):
+    st.header(f"Thread {i}")
+    result_containers.append(st.container())
+
+threads = [WorkerThread(delay) for delay in delays]
+for thread in threads:
+    thread.start()
+thread_lives = [True] * len(threads)
+
+while any(thread_lives):
+    for i, thread in enumerate(threads):
+        if thread_lives[i] and not thread.is_alive():
+            result_containers[i].write(thread.return_value)
+            thread_lives[i] = False
+    time.sleep(0.5)
+
+for thread in threads:
+    thread.join()
+
+st.button("Rerun")
+```
+
+<Cloud name="doc-multithreading-no-st-commands-iterative" height="700px" />
+
+### Option 2: Expose `ScriptRunContext` to the thread
+
+If you want to call Streamlit commands from within your custom threads, you must attach the correct `ScriptRunContext` to the thread.
+
+<Warning>
+
+- This is not officially supported and may change in a future version of Streamlit.
+- This may not work with all Streamlit commands.
+- Ensure custom threads do not outlive the script thread owning the `ScriptRunContext`. Leaking of `ScriptRunContext` may cause security vulnerabilities, fatal errors, or unexpected behavior.
+
+</Warning>
+
+In the following example, a custom thread with `ScriptRunContext` attached can call `st.write` without a warning.
+
+```python
+import streamlit as st
+from streamlit.runtime.scriptrunner import add_script_run_ctx, get_script_run_ctx
+import time
+from threading import Thread
+
+
+class WorkerThread(Thread):
+    def __init__(self, delay, target):
+        super().__init__()
+        self.delay = delay
+        self.target = target
+
+    def run(self):
+        # runs in custom thread, but can call Streamlit APIs
+        start_time = time.time()
+        time.sleep(self.delay)
+        end_time = time.time()
+        self.target.write(f"start: {start_time}, end: {end_time}")
+
+
+delays = [5, 4, 3, 2, 1]
+result_containers = []
+for i, delay in enumerate(delays):
+    st.header(f"Thread {i}")
+    result_containers.append(st.container())
+
+threads = [
+    WorkerThread(delay, container)
+    for delay, container in zip(delays, result_containers)
+]
+for thread in threads:
+    add_script_run_ctx(thread, get_script_run_ctx())
+    thread.start()
+
+for thread in threads:
+    thread.join()
```

Comment:
(Commenting for engineering review later in the week when people are back from the holidays):

Should we be storing the threads in Session State and running a check for threads at the top of the script? And/or disabling widgets when threading? Adding try-except to the Streamlit commands in the threads? Although the script works like this, if a user reruns the app before the page finishes loading, that'd be an issue. Hence this is very fragile, right?